ggml: Add run-time detection of neon, i8mm and sve

eddnjjn opened the pull request:
This patch adds run-time detection of the Arm instructions set features Arm® Neon™, i8mm and sve for Linux and Apple build targets. The run-time detection is enabled for aarch64 builds and done in ggml_init. The data is stored in a global struct instance to be later used by the ggml_cpu_has_* functions.



- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

eddnjjn committed:
ggml: Added run-time detection of neon, i8mm and sve

Adds run-time detection of the Arm instructions set features
neon, i8mm and sve for Linux and Apple build targets.

ggerganov commented on ggml/src/ggml.c:
```diff
@@ -3699,10 +3768,8 @@ struct ggml_context * ggml_init(struct ggml_init_params params) {
-    if (!ggml_sve_cnt_b) {
-        ggml_sve_cnt_b = PR_SVE_VL_LEN_MASK & prctl(PR_SVE_GET_VL);
-    }
+#if defined(__aarch64__)
+    ggml_init_aarch64_features();
 #endif
```
For clarity, this call should be within the `is_first_call` section above

eddnjjn commented:
Thanks for the review @ggerganov . I've rebased the patch and addressed your comment by moving the invocation of ggml_init_aarch64_features to the is_first_call section. I also updated ggml_init_aarch64_features to not check for first invocation since this is done in ggml_init.

Please let me know if you have additional comments.

ggerganov commented on ggml/src/ggml.c:
```diff
@@ -3693,6 +3758,10 @@ struct ggml_context * ggml_init(struct ggml_init_params params) {
             GGML_PRINT_DEBUG("%s: g_state initialized in %f ms\n", __func__, (t_end - t_start)/1000.0f);
         }
 
+#if defined(__aarch64__)
+        ggml_init_aarch64_features();
+#endif
+
```
This looks incorrect because ARM NEON presence is now associated with `__aarch64__`, but this is not always the case AFAIK. For example, here we have support for `__ARM_NEON && !__aarch64__`, such as Raspberry Pi:

https://github.com/ggerganov/llama.cpp/blob/64c6af3195c3cd4aa3328a1282d29cd2635c34c9/ggml/src/ggml-cpu-impl.h#L132-L167

eddnjjn replied:
I've addressed this comment in the update. Non aarch64 Linux builds fallback to using compile time flags for detecting whether or not Neon is enabled.

eddnjjn committed:
ggml: Extend feature detection to include non aarch64 Arm arch

ggerganov commented on ggml/src/ggml.c:
```diff
@@ -37,9 +37,15 @@
 
-#if defined(__ARM_FEATURE_SVE)
-int ggml_sve_cnt_b = 0;
+#if defined(__ARM_ARCH)
+struct ggml_arm_arch_features_type {
+    int has_neon;
+    int has_i8mm;
+    int has_sve;
+    int sve_cnt;
+} ggml_arm_arch_features = {-1, -1, -1, 0};
 #endif
+
```
Move down in the `global data` section, around line 438

eddnjjn replied:
Done!

eddnjjn committed:
ggml: Move definition of ggml_arm_arch_features to the global data section

eddnjjn commented:
I've addressed your latest comment and rebased the patches on latest master. Please let me know if you have additional comments.

ggerganov commented on ggml/include/ggml.h:
```diff
@@ -2507,6 +2507,9 @@ extern "C" {
     GGML_API int ggml_cpu_has_llamafile  (void);
 
+    // get the sve vector length in bytes
+    GGML_API int ggml_cpu_get_sve_cnt(void);
```
This function is better to not be facing the public API. Will merge this for now, but consider making it private.

wtarreau commented:
This series breaks the build on Linux (Ubuntu 20.04 here) due to HWCAP2_I8MM not being known (probably depends on libc or kernel source version):
```
ggml/src/ggml.c: In function "ggml_init_arm_arch_features"
ggml/src/ggml.c:3696:51: error: "HWCAP2_I8MM" undeclared (first use in this function)
 3696 |     ggml_arm_arch_features.has_i8mm = !!(hwcap2 & HWCAP2_I8MM);
      |                                                   ^~~~~~~~~~~
ggml/src/ggml.c:3696:51: note: each undeclared identifier is reported only once for each function it appears in
make: *** [Makefile:1049: ggml/src/ggml.o] Error 1
```
This fixes it:
```
diff --git a/ggml/src/ggml.c b/ggml/src/ggml.c
index fac4466e..854949cb 100644
--- a/ggml/src/ggml.c
+++ b/ggml/src/ggml.c
@@ -3693,7 +3693,9 @@ static void ggml_init_arm_arch_features(void) {
-    uint32_t hwcap2 = getauxval(AT_HWCAP2);
+    uint32_t hwcap2 __attribute__((unused)) = getauxval(AT_HWCAP2);
 
     ggml_arm_arch_features.has_neon = !!(hwcap & HWCAP_ASIMD);
+#if defined(HWCAP2_I8MM)
     ggml_arm_arch_features.has_i8mm = !!(hwcap2 & HWCAP2_I8MM);
+#endif
     ggml_arm_arch_features.has_sve  = !!(hwcap & HWCAP_SVE);
 
 #if defined(__ARM_FEATURE_SVE)
```

But I think it would be cleaner to just define the missing HWCAP* flags before such as this, which is easy to extend to new flags when needed (and works as well):
```
diff --git a/ggml/src/ggml.c b/ggml/src/ggml.c
index fac4466e..f85de4a1 100644
--- a/ggml/src/ggml.c
+++ b/ggml/src/ggml.c
@@ -3687,6 +3687,10 @@ static inline int ggml_up(int n, int m) {
 #include <sys/sysctl.h>
 #endif

+#if !defined(HWCAP2_I8MM)
+#define HWCAP2_I8MM 0
+#endif
+
 static void ggml_init_arm_arch_features(void) {
 #if defined(__linux__) && defined(__aarch64__)
     uint32_t hwcap = getauxval(AT_HWCAP);
```
